test/test.cpp

```diff
@@ -0,0 +1,78 @@
+/*
```

Comment:
Here is an equivalent sample, with _additional more-powerful matching patterns_ than (currently implemented in this PR) in the file test.cpp. We don't have to implement regexp related stuff at all or enhance it in future. It is all there. BTW, it is fast, numbers are below.

```
#include <iostream>
#include <string>
#include <regex>

std::string do_regexp(const std::string & str, const std::string & rstr) {
  std::regex r(rstr);
  std::smatch m;
  std::cout << str << "  [" << rstr << "] ---> " << std::flush;
  auto match = std::regex_search(str, m, r);
  if(match) 
    std::cout << m.str();	// the matched part
  std::cout << '\n';

  return m.str();
}

int main(int argc, char * argv[])
{
  int ct = 1;
  if(argc > 1) ct = atoi(argv[1]);
  std::cout << "Count: " << ct << std::endl;

  if(ct < 1 || ct > 5000) return -1; // too much.. ?
  
  for(int i = 0; i < ct; i++) {
    // next 3: not matching fully, which can be checked for in many ways:
    do_regexp("ab", "a");
    do_regexp("ba", "a");
    do_regexp("bac", "a");

    // exact match..
    do_regexp("ab", "ab");
    
    do_regexp("/foo/bar/baz/blig/fig/blig", "/foo/.*/blig");
    do_regexp("/foo/bar/baz/xlig/fig/blig", "/foo/.*/blig");
    do_regexp("aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaab", "a*a*a*a*a*a*a*a*b");
    do_regexp("aabaabaab", "a.*");
    do_regexp("aabaabaab", "a.*b.*ab");
    do_regexp("aabaabaab", "a.*baab");
    do_regexp("aabaabaab", "aa.*");
    do_regexp("aabaabaab", "aaba.*");
    do_regexp("aabaabqqbaab", "a.*baab");
    do_regexp("aabaabqqbaab", "a.*baab");
    do_regexp("abcdd", ".*d");
    do_regexp("abcdd", ".*d*");
    do_regexp("daaadabadmanda", "da.*da.*da.*");
    do_regexp("mississippi", "m*issip*");
    do_regexp("a big cat", "zz"); // no match

    // Some fun with regexp for non-spaces or spaces:
    // non spaces is a combo which * can't handle..
    do_regexp("a catalog for my bobcat", "\\S+cat$");
    do_regexp("a catalog for my bobcat", "\\s+cat$");
    do_regexp("a catalog for my bob cat", "\\s+cat$");

    // a notion of end of a line:
    do_regexp("a big cat", "a big$");
    do_regexp("a big", "a big$");

    // a notion of begining of a line:
    do_regexp("a big cat", "^a big");  // match
    do_regexp("a big mess", "a big$"); // no match
    

    do_regexp("aabaabqqbaab", "a.*baab");
    do_regexp("abcdd", ".*d");
    do_regexp("abcdd", ".*d.*");

    // Single wildcard
    do_regexp("abc", "a.?c");

    // Special characters
    do_regexp("test.foo[gpu]", "test.foo\\[gpu\\]");
    do_regexp("test.foo[gpu]", "test.foo\\[.*\\]");
    do_regexp("test.foo[gpu]", ".*\\[.*\\]");

    do_regexp("test.foo(gpu)", "test.foo\\(gpu\\)");
    do_regexp("test.foo(gpu)", "test.foo\\(.*\\)");
    do_regexp("test.foo(gpu)", ".*\\(.*\\)");

    do_regexp("test.foog", "test.foo\\[gpu\\]");
    do_regexp("test.foogpu", "test.foo\\[gpu\\]");
    do_regexp("test_foo", "test.foo");
  }    
  return 0;
}
```

Compiled with -O2 with gcc on rome-6 it just took 4 milli seconds for all this stuff in this file. It is pointless to compare performances here -- for deployment of test scripts -- this _4 ms_ or something alike is too little to matter, even if the other approach might be faster:
```
real    0m0.004s
user    0m0.004s
sys     0m0.000s
```
Please also see the additional built-in regexp capabilities that are highlighted above, just to give an example of the powerful library that we should not ignore, and write our own code.

Reply:
You are missing the case for `EXPECT(glob_match("aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaab", "a*a*a*a*a*a*a*a*a*a*a*a*a*b"));`. And, a lot of the regexes are not correct translations(the first three test cases match when they shouldn't match, you are missing escaping `.`, `m*issip*` does not work the same, etc).  You can see the test cases I did for the regex version here:

https://github.com/ROCm/AMDMIGraphX/blob/test-driver-show-time-and-glob-regex/test/test.cpp#L88

I translated each globbing expression to a regex manually, and I ran the cases with rocm clang with -O3.